src/light_data/light_store.rs

```diff
@@ -132,36 +150,116 @@ impl<F: LurkField> LightStore<F> {
+            }
+            Some(Some(expr)) => match (ptr.tag(), expr.clone()) {
+                (ExprTag::Nil, LightExpr::Nil) => {
+                    stack.push(ScalarPtr::from_parts(ExprTag::Sym, *ptr.value()));
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Nil));
+                }
+                (ExprTag::Cons, LightExpr::Cons(x, y)) => {
+                    stack.push(x);
+                    stack.push(y);
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Cons(x, y)));
                 }
-                Some(LightExpr::Nil) => Some(ScalarExpression::Nil),
-                Some(LightExpr::StrNil) => Some(ScalarExpression::Str(String::from(""))),
-                // TODO: StrCons with opaque contents breaks this
-                Some(LightExpr::StrCons(_, _)) => {
-                    self.insert_scalar_string(*ptr, &mut store)?;
-                    continue;
+                (ExprTag::Comm, LightExpr::Comm(f, x)) => {
+                    stack.push(x);
+                    store.insert_scalar_expression(ptr, Some(ScalarExpression::Comm(f, x)));
                 }
-                Some(LightExpr::SymNil) => Some(ScalarExpression::Sym(Sym::root())),
-                // TODO: SymCons with opaque contents breaks this
-                Some(LightExpr::SymCons(_, _)) => {
-                    self.insert_scalar_symbol(*ptr, &mut store)?;
-                    continue;
+                (ExprTag::Sym, LightExpr::SymCons(_, _)) => {
+                    self.insert_scalar_symbol(ptr, store)?;
                 }
-            };
-            store.insert_scalar_expression(*ptr, se);
+                (ExprTag::Str, LightExpr::StrCons(_, _)) => {
+                    self.insert_scalar_string(ptr, store)?;
+                }
+                (tag, _) => {
+                    return Err(anyhow!(
+                        "Invalid pair of tag and LightExpr: ({tag}, {expr})"
+                    ))
+                }
+            },
+        };
+        Ok(())
+    }
+
+    /// Eagerly traverses the LightStore starting out from a ScalarPtr, adding
+    /// pointers and their respective expressions to a target ScalarStore. When
+    /// handling non-immediate pointers, their corresponding expressions might
+    /// add more pointers to be visited to a stack.
+    ///
+    /// TODO: add a cycle detection logic
+    fn intern_ptr_data(
+        &self,
+        ptr0: ScalarPtr<F>,
+        store: &mut ScalarStore<F>,
+    ) -> anyhow::Result<()> {
+        let mut stack = Vec::new();
+        stack.push(ptr0);
+        while let Some(ptr) = stack.pop() {
+            if store.get_expr(&ptr).is_some() {
+                continue;
+            }
+            if ptr.is_immediate() {
+                self.intern_immediate(ptr, store)?;
+            } else {
+                self.intern_non_immediate(ptr, store, &mut stack)?;
+            }
```

Comment:
This creates a lot of coupling between these methods (and also the ones the interning functions call). I think using this predicate-based dispatch here is brittle, and in fact it indirectly leads to some oversights already. Given that, I think it would be better to structure this in such a way that we directly rely on the type system and have fewer invariants the programmer needs to 'just keep track' of. I'll make some suggestions as to how I think this could be done.

If we were to leave this code structured as it is, I would at least add assertions in `intern_immediate` and `intern_non_immediate` checking to make sure this dispatch has happened correctly — since those functions depend on not being called with the wrong kind of values.

Reply:
I'm moving `is_immediate` to a `is_ptr_immediate` in the context of `LightStore`